Support generating log methods in nested classes

maryamariyan opened the pull request:
- [x] Add test for non static nested case
- [x] Support log methods in nested records/structs/classes

Fixes #52301

maryamariyan committed:
Support log methods in nested classes

maryamariyan committed:
Remove message for nested error

maryamariyan commented:
Notes for review:

- Might be easier to review by commit (second commit is just a resx change)
- Review tests first, then Parser, then Emitter code changes

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorEmitterTests.cs:
```diff
@@ -80,6 +80,32 @@ internal static partial class TestWithDynamicLogLevel
+{
+    namespace NestedNamespace
+    {
+        internal static partial class NestedClassTestsExtensions<T1> where T1 : Class1
```
Can you also add a test where the outer class isn't `static`?

And also a test for where the outer class isn't `partial`, or at least verify the behavior manually? To ensure the user gets a decent error experience so they know what they did wrong.

eerhardt replied:
It might also be good to test with a `struct` outer type as well.

maryamariyan replied:
> And also a test for where the outer class isn't partial, or at least verify the behavior manually? To ensure the user gets a decent error experience so they know what they did wrong.

For something like
```c#
    class ClassNotPartial
    {
        internal static partial class NestedClass
        {
            [LoggerMessage(EventId = 10, Level = LogLevel.Debug, Message = "M10")]
            public static partial void M10(ILogger logger);
        }
    }
```

We'd get
```
error CS0260: Missing partial modifier on declaration of type 'ClassNotPartial'; another partial declaration of this type exists 
```
I think the error message is clear enough.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -65,29 +65,60 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
 
             private void GenType(LoggerClass lc)
             {
+                int indentationSize = 0;
```
Why do we need to keep track of both `int indentationSize` and `string nestedIndentation`? Couldn't we make due with just the string, and append / trim 4 spaces as necessary?

maryamariyan replied:
removed `indentationSize`

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -571,6 +586,7 @@ internal class LoggerClass
             public string Namespace = string.Empty;
             public string Name = string.Empty;
             public string Constraints = string.Empty;
+            public LoggerClass? ParentClass = null;
```
I don't think you need the `= null` part. It is redundant.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -331,28 +331,25 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
+                                                potentialNamespaceParent = potentialNamespaceParent.Parent;
                                             }
-                                            else
+                                            if (potentialNamespaceParent != null && potentialNamespaceParent as NamespaceDeclarationSyntax != null)
```
```suggestion
                                            if (potentialNamespaceParent != null && potentialNamespaceParent is not NamespaceDeclarationSyntax)
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -331,28 +331,25 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
-                                            if (ns == null)
+                                            NamespaceDeclarationSyntax? namespaceParent = null;
+                                            SyntaxNode? potentialNamespaceParent = classDec.Parent;
+                                            while (potentialNamespaceParent != null && potentialNamespaceParent as NamespaceDeclarationSyntax == null)
```
```suggestion
                                            while (potentialNamespaceParent != null && potentialNamespaceParent is not NamespaceDeclarationSyntax)
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -364,8 +361,26 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
                                             };
 
+                                            LoggerClass currentLoggerClass = lc;
+                                            var parentLoggerClass = (classDec.Parent as ClassDeclarationSyntax);
```
Does `ClassDeclarationSyntax` include `struct` or `record`?

maryamariyan replied:
No I think we need `StructDeclarationSyntax` for that

maryamariyan replied:
New commit adds support for logger methods nested inside structs, classes and records

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -65,29 +65,60 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
+                // loop until you find top level nested class
+                while (parent != null)
+                {
+                    parentClasses.Push($@"partial class {parent?.Name + " " + parent?.Constraints}");
```
`struct` / `record`?

maryamariyan replied:
log methods in structs are already not supported by the logging source generator. maybe it can be covered later.

e.g. with:

```c#

    internal partial struct MyStruct
    {
        private ILogger _logger;
        
        public MyStruct(ILogger logger) { _logger = logger; }

        [LoggerMessage(EventId = 9, Level = LogLevel.Debug, Message = "M9")]
        public partial void M9();
    }
```

we get: 
```
 error CS8795: Partial method 'NestedClass.M9()' must have an implementation part because it has accessibility modifiers. 
```

eerhardt replied:
> log methods in structs

I wasn't referring to log methods in structs. But log methods in a nested class of a struct / record. ex

```C#
public struct Customer
{
    private static class Logger
    {
        [LoggerMessage(EventId = 9, Level = LogLevel.Debug, Message = "M9")]
        public static void M9();
    }
}
```

```C#
public record Customer(string Name, string Address)
{
    private static class Logger
    {
        [LoggerMessage(EventId = 9, Level = LogLevel.Debug, Message = "M9")]
        public static void M9();
    }
}
```

maryamariyan replied:
Added above use cases as tests. Also added support for adding logger methods directly in structs

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -65,29 +65,60 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
                 }
 
+                LoggerClass parent = lc.ParentClass;
+                Stack<string> parentClasses = new Stack<string>();
```
Since you don't really "push" and "pop" things on and off this stack, you could just use a `List<string>` and then walk over it in the reverse order `for (int i = parentClasses.Length - 1; i >= 0; i--)`.

maryamariyan committed:
Apply some PR feedback

maryamariyan committed:
- Remove variable indentationSize
- Add test for non static nested case
- Add support for log method directly in structs
- Support log methods in nested records/structs

maryamariyan committed:
- Cleanup tests
- Undo support for loggers directrly under struct

maryamariyan commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -28,7 +28,7 @@ public Parser(Compilation compilation, Action<Diagnostic> reportDiagnostic, Canc
 
             /// <summary>
-            /// Gets the set of logging classes containing methods to output.
+            /// Gets the set of logging classes or structs containing methods to output.
```
```suggestion
            /// Gets the set of logging classes containing methods to output.
```

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -331,28 +331,24 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
+                                            if (potentialNamespaceParent != null && potentialNamespaceParent is NamespaceDeclarationSyntax namespaceParent)
                                             {
-                                                nspace = ns.Name.ToString();
+                                                namespaceParent = potentialNamespaceParent as NamespaceDeclarationSyntax;
```
This line is redundant with the above `potentialNamespaceParent is NamespaceDeclarationSyntax namespaceParent`. I believe this line can just be deleted.

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs:
```diff
@@ -331,28 +331,24 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
+                                                potentialNamespaceParent = potentialNamespaceParent.Parent;
                                             }
-                                            else
+                                            if (potentialNamespaceParent != null && potentialNamespaceParent is NamespaceDeclarationSyntax namespaceParent)
```
You don't need to check for `null` at the beginning. You can just say:

`if (potentialNamespaceParent is NamespaceDeclarationSyntax namespaceParent)`

eerhardt commented on src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Emitter.cs:
```diff
@@ -65,29 +65,57 @@ private static bool UseLoggerMessageDefine(LoggerMethod lm)
+                // loop until you find top level nested class
+                while (parent != null)
+                {
+                    parentClasses.Add($"partial {parent?.Keyword} {parent?.Name} {parent?.Constraints}");
```
Why do you need all these `?` on `parent`? You know it is non-null (from the loop condition), so you should be able to remove them.

eerhardt reviewed (approved):
LGTM. Thanks for the good work here!

maryamariyan committed:
Address PR feedback

maryamariyan commented:
/backport to release/5.0-preview6

maryamariyan commented:
/backport to release/6.0-preview6

maryamariyan commented:
CI results from later commits in the PR: [buildId 1188639](https://dev.azure.com/dnceng/public/_build/results?buildId=1188639&view=results)

More CI related to the PR here:
- [buildId 1189154](https://dev.azure.com/dnceng/public/_build/results?buildId=1189154&view=results)
- [buildId 1189108](https://dev.azure.com/dnceng/public/_build/results?buildId=1189108&view=results)
- [buildId 1189215](https://dev.azure.com/dnceng/public/_build/results?buildId=1189215&view=results)
